src/python/utils/diff_statistics.py

```diff
@@ -0,0 +1,198 @@
+from pathlib import Path
+from typing import Union, Set
+
+from loguru import logger
+
+from src.python.repo import Repository
+
+
+def get_diff(
+    repo: Repository, left_sha: str, right_sha: str, diff_log_file: Union[None, Path]
+) -> str:
+    """
+    Computes the git diff between two commits.
+    Args:
+        repo (Repository): The repository object.
+        left_sha (str): The left sha.
+        right_sha (str): The right sha.
+    Returns:
+        str: A string containing the diff result.
+    """
+    command = f"git diff {left_sha} {right_sha}"
+    stdout, _ = repo.run_command(command)
+    return stdout
+
+
+def get_diff_files(repo: Repository, left_sha: str, right_sha: str) -> set:
+    """
+    Computes the set of files that are different between two commits using git diff.
+    Args:
+        repo (Repository): The repository object.
+        left_sha (str): The left sha.
+        right_sha (str): The right sha.
+    Returns:
+        set: A set containing the diff result.
+    """
+    # Using git diff to compare the two SHAs
+    command = f"git diff --name-only {left_sha} {right_sha}"
+    stdout, _ = repo.run_command(command)
+    return set(stdout.split("\n")) if stdout else set()
+
+
+def get_diff_hunks(repo: Repository, left_sha: str, right_sha: str) -> int:
+    """
+    Compute the number of hunks that are different between two commits using git diff.
+    Args:
+        repo (Repository): The repository object.
+        left_sha (str): The left sha.
+        right_sha (str): The right sha.
+    Returns:
+        int: The number of hunks that are different between the two branches.
+    """
+    diff, _ = repo.run_command(
+        f"git diff --unified=0 {left_sha} {right_sha} | grep -c '^@@'"
+    )
+    return int(diff)
+
+
+def get_diff_files_merge(
+    repo: Repository,
+    left_sha: str,
+    right_sha: str,
+) -> Set[str]:
+    """
+    Computes the intersection of files that are different between a three-way merge using git diff.
+    Args:
+        repo (Repository): The repository object.
+        merge_idx (str): The merge index, such as 42-123.
+        repo_slug (str): The repository slug.
+        left_sha (str): The left sha.
+        right_sha (str): The right sha.
+        cache_dir (Path): The path to the cache directory.
+    Returns:
+        Set[str]: A set containing the diff result.
+    """
+    command = f"git merge-base {left_sha} {right_sha}"
+    base_sha = repo.run_command(command)[0].strip()
+    left_right_files = get_diff_files(repo, left_sha, right_sha)
+    base_right_files = get_diff_files(repo, base_sha, right_sha)
+    base_left_files = get_diff_files(repo, base_sha, left_sha)
+
+    # Check that at least one java file is contained in all 3 diffs
+    common_files = left_right_files & base_right_files & base_left_files
```

Comment:
I'm not sure whether this is right without a clarification of the function documentation.
But I wouldn't have expected left_right_files to figure here.